File: backend/src/services/template_service.py
"""Service layer for template operations."""

from sqlalchemy.orm import Session
from typing import Optional
from fastapi import HTTPException, status, UploadFile
from ..models.template import Template, TemplateType
from ..schemas.template import TemplateCreate, TemplateUpdate
from ..utils.document_processor import DocumentProcessor
from ..config import settings
import os


class TemplateService:
    """Service for template CRUD operations."""
    
    @staticmethod
    def create_template(db: Session, template_data: TemplateCreate, created_by: int) -> Template:
        """
        Create a new template.
        
        Args:
            db: Database session
            template_data: Template creation data
            created_by: User ID creating the template
            
        Returns:
            Created template
        """
        # Validate markdown content
        if not DocumentProcessor.validate_markdown(template_data.markdown_content):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid Markdown content"
            )
        
        template = Template(
            name=template_data.name,
            description=template_data.description,
            template_type=template_data.template_type,  # Pass the string directly, SQLAlchemy will handle it
            markdown_content=template_data.markdown_content,
            original_filename=template_data.original_filename,
            original_file_path=template_data.original_file_path,
            created_by=created_by
        )
        
        db.add(template)
        db.commit()
        db.refresh(template)
        
        return template
    
    @staticmethod
    def get_template(db: Session, template_id: int) -> Optional[Template]:
        """
        Get template by ID.
        
        Args:
            db: Database session
            template_id: Template ID
            
        Returns:
            Template if found, None otherwise
        """
        return db.query(Template).filter(
            Template.id == template_id,
            Template.is_active == True
        ).first()
    
    @staticmethod
    def list_templates(
        db: Session,
        skip: int = 0,
        limit: int = 100,
        search: Optional[str] = None
    ) -> tuple[list[Template], int]:
        """
        List templates with pagination and search.
        
        Args:
            db: Database session
            skip: Number of records to skip
            limit: Maximum number of records to return
            search: Optional search term for name
            
        Returns:
            Tuple of (templates list, total count)
        """
        query = db.query(Template).filter(Template.is_active == True)
        
        if search:
            query = query.filter(Template.name.ilike(f"%{search}%"))
        
        total = query.count()
        templates = query.order_by(Template.created_at.desc()).offset(skip).limit(limit).all()
        
        return templates, total
    
    @staticmethod
    def update_template(
        db: Session,
        template_id: int,
        template_data: TemplateUpdate
    ) -> Optional[Template]:
        """
        Update template.
        
        Args:
            db: Database session
            template_id: Template ID
            template_data: Update data
            
        Returns:
            Updated template if found, None otherwise
        """
        template = TemplateService.get_template(db, template_id)
        if not template:
            return None
        
        update_data = template_data.model_dump(exclude_unset=True)
        
        # Validate markdown if being updated
        if 'markdown_content' in update_data:
            if not DocumentProcessor.validate_markdown(update_data['markdown_content']):
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="Invalid Markdown content"
                )
        
        for field, value in update_data.items():
            setattr(template, field, value)
        
        db.commit()
        db.refresh(template)
        
        return template
    
    @staticmethod
    def delete_template(db: Session, template_id: int) -> bool:
        """
        Soft delete template.
        
        Args:
            db: Database session
            template_id: Template ID
            
        Returns:
            True if deleted, False if not found
        """
        template = TemplateService.get_template(db, template_id)
        if not template:
            return False
        
        template.is_active = False
        db.commit()
        
        return True
    
    @staticmethod
    async def process_uploaded_file(
        file: UploadFile,
        created_by: int,
        db: Session = None,
        template_name: str = None
    ) -> dict:
        """
        Process uploaded file and convert to markdown.

        Args:
            file: Uploaded file
            created_by: User ID uploading the file
            db: Database session (optional, for fetching user info)
            template_name: Template name for markdown file naming (optional)

        Returns:
            Dictionary with file info and markdown content
        """
        # Determine file type
        file_type = DocumentProcessor.get_file_type(file.filename)
        if not file_type:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Unsupported file type. Supported: .docx, .pdf, .txt, .jpg, .png, .tiff"
            )

        # Read file content
        content = await file.read()

        # Save file to storage
        upload_dir = settings.upload_dir
        file_path = DocumentProcessor.save_uploaded_file(content, file.filename, upload_dir)

        # Convert to markdown based on file type
        try:
            if file_type == 'word':
                markdown_content = DocumentProcessor.word_to_markdown(file_path)
            elif file_type == 'pdf':
                markdown_content = DocumentProcessor.pdf_to_markdown(file_path)
            elif file_type == 'text':
                markdown_content = DocumentProcessor.text_to_markdown(file_path)
            elif file_type == 'image':
                # For images, we would call AWS Textract here
                # For now, return placeholder
                markdown_content = "# OCR Processing Required\n\nPlease process this image with AWS Textract."
            else:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="Unsupported file type"
                )

            # Save markdown file to document_uploads if Word, PDF, or Text
            markdown_file_path = None
            if file_type in ['word', 'pdf', 'text'] and db and template_name:
                # Get username from database
                from ..models.user import User
                user = db.query(User).filter(User.id == created_by).first()
                username = user.username if user else f"user_{created_by}"

                # Save markdown file
                markdown_file_path = DocumentProcessor.save_markdown_file(
                    markdown_content,
                    template_name,
                    username
                )

        except Exception as e:
            # Clean up file if processing fails
            if os.path.exists(file_path):
                os.remove(file_path)
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=f"Error processing file: {str(e)}"
            )

        return {
            "filename": file.filename,
            "file_path": file_path,
            "file_type": file_type,
            "markdown_content": markdown_content,
            "markdown_file_path": markdown_file_path
        }
    
    @staticmethod
    def get_template_identifiers(db: Session, template_id: int) -> Optional[list[str]]:
        """
        Get all identifiers from a template.
        
        Args:
            db: Database session
            template_id: Template ID
            
        Returns:
            List of identifiers if template found, None otherwise
        """
        template = TemplateService.get_template(db, template_id)
        if not template:
            return None
        
        return template.extract_identifiers()

File: go-backend/internal/services/template_service.go
package services

import (
	"errors"
	"strings"
	"time"

	"github.com/estate-docs/go-backend/internal/models"
	"gorm.io/gorm"
)

type TemplateService struct {
	DB *gorm.DB
}

func NewTemplateService(db *gorm.DB) *TemplateService {
	return &TemplateService{DB: db}
}

type TemplateCreateInput struct {
	Name             string  `json:"name"`
	Description      *string `json:"description"`
	TemplateType     string  `json:"template_type"`
	MarkdownContent  string  `json:"markdown_content"`
	OriginalFilename *string `json:"original_filename"`
	OriginalFilePath *string `json:"original_file_path"`
}

type TemplateUpdateInput struct {
	Name            *string `json:"name"`
	Description     *string `json:"description"`
	MarkdownContent *string `json:"markdown_content"`
}

func (s *TemplateService) Create(input TemplateCreateInput, userID int) (*models.Template, error) {
	t := models.Template{
		Name:             input.Name,
		Description:      input.Description,
		TemplateType:     input.TemplateType,
		MarkdownContent:  input.MarkdownContent,
		OriginalFilename: input.OriginalFilename,
		OriginalFilePath: input.OriginalFilePath,
		CreatedBy:        &userID,
		IsActive:         true,
		CreatedAt:        time.Now().UTC(),
		UpdatedAt:        time.Now().UTC(),
	}

	// Extract and store identifiers
	ids := t.ExtractIdentifiers()
	if len(ids) > 0 {
		joined := strings.Join(ids, ",")
		t.Identifiers = &joined
	}

	if err := s.DB.Create(&t).Error; err != nil {
		return nil, err
	}
	return &t, nil
}

func (s *TemplateService) List(skip, limit int, search *string) ([]models.Template, int64, error) {
	var templates []models.Template
	var total int64

	q := s.DB.Model(&models.Template{}).Where("is_active = true")
	if search != nil && *search != "" {
		q = q.Where("name ILIKE ?", "%"+*search+"%")
	}
	q.Count(&total)

	if err := q.Order("id DESC").Offset(skip).Limit(limit).Find(&templates).Error; err != nil {
		return nil, 0, err
	}
	return templates, total, nil
}

func (s *TemplateService) GetByID(id int) (*models.Template, error) {
	var t models.Template
	if err := s.DB.Where("is_active = true").First(&t, id).Error; err != nil {
		return nil, err
	}
	return &t, nil
}

func (s *TemplateService) Update(id int, input TemplateUpdateInput) (*models.Template, error) {
	var t models.Template
	if err := s.DB.First(&t, id).Error; err != nil {
		return nil, errors.New("template not found")
	}

	if input.Name != nil {
		t.Name = *input.Name
	}
	if input.Description != nil {
		t.Description = input.Description
	}
	if input.MarkdownContent != nil {
		t.MarkdownContent = *input.MarkdownContent
		ids := t.ExtractIdentifiers()
		if len(ids) > 0 {
			joined := strings.Join(ids, ",")
			t.Identifiers = &joined
		}
	}
	t.UpdatedAt = time.Now().UTC()

	if err := s.DB.Save(&t).Error; err != nil {
		return nil, err
	}
	return &t, nil
}

func (s *TemplateService) Delete(id int) error {
	var t models.Template
	if err := s.DB.First(&t, id).Error; err != nil {
		return errors.New("template not found")
	}
	t.IsActive = false
	t.UpdatedAt = time.Now().UTC()
	return s.DB.Save(&t).Error
}

func (s *TemplateService) Duplicate(id, userID int) (*models.Template, error) {
	var orig models.Template
	if err := s.DB.First(&orig, id).Error; err != nil {
		return nil, errors.New("template not found")
	}

	dup := models.Template{
		Name:             orig.Name + " (Copy)",
		Description:      orig.Description,
		TemplateType:     orig.TemplateType,
		MarkdownContent:  orig.MarkdownContent,
		Identifiers:      orig.Identifiers,
		OriginalFilename: orig.OriginalFilename,
		OriginalFilePath: orig.OriginalFilePath,
		CreatedBy:        &userID,
		IsActive:         true,
		CreatedAt:        time.Now().UTC(),
		UpdatedAt:        time.Now().UTC(),
	}

	if err := s.DB.Create(&dup).Error; err != nil {
		return nil, err
	}
	return &dup, nil
}

func (s *TemplateService) GetIdentifiers(id int) ([]string, error) {
	var t models.Template
	if err := s.DB.First(&t, id).Error; err != nil {
		return nil, errors.New("template not found")
	}
	return t.ExtractIdentifiers(), nil
}
